File: plugins/plugin-auto-trader/rust/src/actions/analyze_performance.rs
use async_trait::async_trait;
use serde_json::{json, Value};

use crate::service::TradingService;
use crate::{Action, ActionExample, ActionResult};

pub struct AnalyzePerformanceAction;

#[async_trait]
impl Action for AnalyzePerformanceAction {
    fn name(&self) -> &str {
        "ANALYZE_PERFORMANCE"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "PERFORMANCE_REPORT",
            "TRADING_STATS",
            "SHOW_PERFORMANCE",
            "GET_STATS",
        ]
    }

    fn description(&self) -> &str {
        "Analyze trading performance and return a detailed report."
    }

    async fn validate(&self, _message: &Value, _state: &Value) -> bool {
        true
    }

    async fn handler(
        &self,
        _message: &Value,
        _state: &Value,
        service: Option<&TradingService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "TradingService is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        let report = svc.analyze_performance().await;

        let text = format!(
            "Performance Report:\n  Total Trades: {}\n  Winning: {} | Losing: {}\n  Win Rate: {:.1}%\n  Total PnL: ${:.2} ({:+.2}%)\n  Avg Win: ${:.2} | Avg Loss: ${:.2}\n  Max Drawdown: {:.2}%\n  Sharpe Ratio: {:.2}",
            report.total_trades,
            report.winning_trades,
            report.losing_trades,
            report.win_rate * 100.0,
            report.total_pnl,
            report.total_pnl_pct,
            report.avg_win,
            report.avg_loss,
            report.max_drawdown * 100.0,
            report.sharpe_ratio,
        );

        ActionResult {
            success: true,
            text,
            data: Some(json!(report)),
            error: None,
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "Show my trading performance".to_string(),
            agent_response: "Performance Report: ...".to_string(),
        }]
    }
}

File: plugins/plugin-auto-trader/python/elizaos_plugin_auto_trader/actions/analyze_performance.py
from __future__ import annotations

from dataclasses import dataclass

from elizaos_plugin_auto_trader.actions.common import ActionResult, Message
from elizaos_plugin_auto_trader.service import TradingService


@dataclass
class AnalyzePerformanceAction:
    @property
    def name(self) -> str:
        return "ANALYZE_PERFORMANCE"

    async def validate(self, _message: Message, _state: dict) -> bool:  # type: ignore[type-arg]
        return True

    async def handler(
        self,
        _message: Message,
        _state: dict,  # type: ignore[type-arg]
        service: TradingService | None = None,
    ) -> ActionResult:
        if service is None:
            return ActionResult(False, "TradingService is not available.", "missing_service")

        report = await service.analyze_performance()

        text = (
            f"Performance Report:\n"
            f"  Total Trades: {report.total_trades}\n"
            f"  Winning: {report.winning_trades} | Losing: {report.losing_trades}\n"
            f"  Win Rate: {report.win_rate * 100:.1f}%\n"
            f"  Total PnL: ${report.total_pnl:.2f} ({report.total_pnl_pct:+.2f}%)\n"
            f"  Avg Win: ${report.avg_win:.2f} | Avg Loss: ${report.avg_loss:.2f}\n"
            f"  Max Drawdown: {report.max_drawdown * 100:.2f}%\n"
            f"  Sharpe Ratio: {report.sharpe_ratio:.2f}"
        )

        return ActionResult(True, text, data=report)
